packages/components-library/src/components/ui/Pagination.vue

```diff
@@ -0,0 +1,287 @@
+      const firstPage = pages[0]
+      if (firstPage === 2) { pages.unshift(1) }
+
+      // The same with the end part:
+      const lastPage = pages[pages.length - 1]
+      if (lastPage === (this.pageCount - 1)) { pages.push(this.pageCount) }
+
+      return pages
+    }
+  },
+  data: function () {
+    return {
+      val: { size: 20, page: 1, loading: false, count: 0 }
+    }
+  },
+  methods: {
+    async fetchData (page) {
+      // Omit the Pagination method when you're using multiple
+      // Pagination components with the same data source.
+      if (!this.method) { return }
+      page = parseInt(page, 10)
+
+      this.updateValue({ loading: true, page })
+
+      const { count } = await this.method()
+      this.updateValue({ loading: false, count, page })
+
+      this.val = { ...this.value, ...{ loading: false, count, page } }
+    },
+    navigate (page) {
+      if (!this.useRouter) { return this.fetchData(page) }
+
+      if (this.$route.query.page) {
+        if (
+          parseInt(this.$route.query.page, 10) !== page ||
+          parseInt(this.$route.query.size, 10) !== this.value.size
+        ) {
+          // (!) Merge with existing query; e.g. bookmarks
+          this.$router.push({
+            path: this.$route.path,
+            query: { ...this.$route.query, page, size: this.value.size }
+          })
+        }
+      } else {
+        this.$router.replace({
+          path: this.$route.path,
+          query: { ...this.$route.query, page, size: this.value.size }
+        })
+      }
+    },
+    updateValue (value) {
+      this.val = { ...this.val, ...this.value, ...value }
+      this.$emit('input', { ...this.value, ...this.val, ...value })
+    }
+  },
+  props: {
+    /**
+     * Add css classes to pagination buttons.
+     */
+    styles: {
+      type: Object,
+      required: false,
+      default: () => ({})
+    },
+    /**
+     * The method to retrieve items with.
+     * @returns {Object} {count: Number}
+     */
+    method: {
+      type: Function,
+      required: false
+    },
+    /**
+     * Amount of navigational page buttons around the current page button.
+     */
+    midPageSize: {
+      type: Number,
+      default: () => 4
+    },
+    /**
+     * Whether to integrate with Vue-router.
+     */
+    useRouter: {
+      type: Boolean,
+      default: () => true
+    },
+    /**
+     * The internal pagination model
+     * @model
+     */
+    value: {
+      type: Object,
+      required: true
+    }
+  },
+  watch: {
+    '$route.query': {
+      handler: function (query) {
+        if (!this.method || !this.useRouter) { return }
+
+        if (query.page) {
+          this.fetchData(parseInt(query.page, 10))
+        } else {
+          this.navigate(1)
+        }
+      },
+      deep: true,
+      immediate: true
+    },
+    value: {
+      handler: function (newVal) {
+        this.val = newVal
+      },
+      deep: true
+    },
+    'value.size' () {
```

Comment:
why do we need to watch value.size when value is already watched ( with deep = true) ?

Reply:
This can be done in one handler indeed. Checking for changed size now in the value watcher.